Applications/Spire/Source/Ui/DecimalBox.cpp

```diff
@@ -380,6 +356,108 @@ struct DecimalBox::DecimalToTextModel : TextModel {
   }
 };
 
+struct DecimalBox::StepButton : QWidget {
+  using PressSignal = Signal<void()>;
+  mutable PressSignal m_press_signal;
+  FocusObserver m_focus_observer;
+  int m_repeat_delay_timer_id;
+  int m_repeat_timer_id;
+
+  StepButton(QImage icon, QWidget& parent)
+      : QWidget(&parent),
+        m_focus_observer(parent),
+        m_repeat_delay_timer_id(-1),
+        m_repeat_timer_id(-1) {
+    auto button_icon = new Icon(std::move(icon));
+    auto box = new Box(button_icon);
+    proxy_style(*this, *box);
+    link(*this, *button_icon);
+    enclose(*this, *box);
+    update_style(*this, [] (auto& style) {
+      style.get(Any() > is_a<Icon>()).
+        set(BackgroundColor(QColor(0xFFFFFF))).
+        set(Fill(QColor(0x333333)));
+      style.get(Hover() > is_a<Icon>()).
+        set(BackgroundColor(QColor(0xEBEBEB))).
+        set(Fill(QColor(0x4B23A0)));
+      style.get(Press() > is_a<Icon>()).
+        set(BackgroundColor(QColor(0xEBEBEB))).
+        set(Fill(QColor(0x7E71B8)));
+      style.get(Disabled() > is_a<Icon>()).
+        set(BackgroundColor(QColor(Qt::transparent))).
+        set(Fill(QColor(0xC8C8C8)));
+    });
+    setFocusPolicy(Qt::NoFocus);
+    setFixedSize(BUTTON_SIZE());
+    m_focus_observer.connect_state_signal(
+      std::bind_front(&StepButton::on_focus, this));
+  }
+
+  connection connect_press_signal(const PressSignal::slot_type& slot) const {
+    return m_press_signal.connect(slot);
+  }
+
+  void mousePressEvent(QMouseEvent* event) override {
+    if(QWidget* parent = parentWidget()) {
+      parent->setFocus();
+    }
+    if(event->button() == Qt::LeftButton) {
+      event->accept();
+      match(*this, Press());
+      m_press_signal();
+      m_repeat_delay_timer_id = startTimer(REPEAT_DELAY_TIMEOUT_MS);
+    } else {
+      QWidget::mousePressEvent(event);
+    }
+  }
+
+  void mouseReleaseEvent(QMouseEvent* event) override {
+    if(event->button() == Qt::LeftButton) {
+      event->accept();
+      reset();
+    } else {
+      QWidget::mouseReleaseEvent(event);
+    }
+  }
+
+  void timerEvent(QTimerEvent* event) override {
+    if(event->timerId() == m_repeat_delay_timer_id) {
+      killTimer(m_repeat_delay_timer_id);
+      m_repeat_delay_timer_id = -1;
+      m_repeat_timer_id = startTimer(REPEAT_TIMEOUT_MS);
+    } else if(event->timerId() == m_repeat_timer_id) {
+      if(isEnabled() && isVisible() &&
+          rect().contains(mapFromGlobal(QCursor::pos()))) {
+        m_press_signal();
+      } else {
+        reset();
+      }
+    } else {
+      QWidget::timerEvent(event);
+    }
+  }
+
+  void reset() {
+    if(is_match(*this, Press())) {
+      unmatch(*this, Press());
```

Comment:
Unmatch unconditionally, it's faster than checking for a match and then removing it.